### ISSUE TITLE
Transferring Face Recognition model to Pytorch. Inconsistent layer outputs.

### ISSUE THREAD

**Issue by redplanet543** [davisking/dlib#1732]
I am trying to transfer the dlib_face_recognition_resnet_model_v1 model to Pytorch. I have transferred all weights and biases for the convolution and affine layers, but when I check the output of the first convolution (layer 130) the tensors I get from Pytorch and dlib do not match in values.

Does this mean that dlib and Pytorch perform convolution differently? I noticed there are zeroes in the output I get from dlib and that seems odd to me.

For dlib I use the network defined in http://dlib.net/dnn_face_recognition_ex.cpp.html, for Pytorch I built my own and copied the weights.

* **Version**: 19.17
* **Where did you get dlib**: github repo
* **Platform**: Ubuntu 18.04.1 LTS
* **Compiler**: g++ 6.5.0

**Comment by ksachdeva** [comment]
I have done this exercise for tensorflow/keras. Please see it here - https://github.com/ksachdeva/dlib-to-tf-keras-converter.

There are few reasons why there is a difference -

a) If you used the process which first convert .dat to .xml then there is a loss of precision in the weights. 

b) The linear algebra libraries used by dlib and tensorflow/pytorch will be different and they contribute to the difference as well.

c) The target CPU/GPU difference also will result in different embeddings.

You can try to run a dlib application on two different machines and you will see the difference in the generated embeddings

Hope this helps

**Comment by timesler** [comment]
@redplanet543 you can try the following repo. It contains facial recognition resnets converted for pytorch:

[https://github.com/timesler/facenet-pytorch](https://github.com/timesler/facenet-pytorch)

**Comment by redionxhepa** [comment]
@redplanet543 Can you please share the Pytorch code how you did the conversion ? Thanks in advance !
